### ISSUE TITLE
Proposal for singleton references and operators

### ISSUE THREAD

**Issue by shadaj** [hydro-project/hydroflow#1065]
Right now, operators like `fold` and `reduce` take in a stream of values and return another stream. This causes two problems:
1) the type information that there is only a single result of the fold is lost
2) using the result of the fold in UDFs requires a `cross_join`, which results in **unnecessary cloning** and an awkward implementation

For example, the current surface syntax to filter elements from one stream based on the sum of another looks like:
```rust
sum_of_stream2 = stream2 -> fold(|| 0, |acc, cur| *acc += cur);
filtered_stream1 = cross_join() -> filter(|(value, sum)| value < sum);
stream1 -> [0] filtered_stream1;
sum_of_stream2 -> [1] filtered_stream2;
```

I propose that we introduce a new concept of singletons, which are emitted by certain operators (the ones which currently emit a stream with a single value). Singletons can be directly referenced from within UDFs using a special syntax (placing a `#` before the identifier for the singleton). Under the hood, singletons would be compiled to a standard Rust variable, which would allow them to be moved or borrowed into UDFs.

For example, the above example could be written as:
```rust
sum_of_stream2 = stream2 -> fold(|| 0, |acc, cur| *acc += cur);
filtered_stream1 = stream1 -> filter(|value| value < #sum_of_stream2);
```

or instead of copying the sum, we can also just reference it
```rust
sum_of_stream2 = stream2 -> fold(|| 0, |acc, cur| *acc += cur);
filtered_stream1 = stream1 -> filter(|value| value < &#sum_of_stream2);
```

Because we use a special `#` syntax to reference singletons, the UDFs can be safely identified to analyze their dependencies and build out the regular Hydroflow compute graph.

**Comment by MingweiSamuel** [comment]
1. Should singletons need to be `Copy`? `Clone`? Or allow `&`? (Or allow `&mut` and get panics?) -- Internally will likely need to use a `RefCell`
2. Do we want to differentiate between pipeline names and singleton names somehow?

**Comment by shadaj** [comment]
1. I think we shouldn't restrict Copy/Clone, but it could be reasonable to only allow `&` (and require explicit wrapping in a `RefCell` if mutability is needed)
2. Maybe not the identifiers, but we'll need some custom syntax for statements creating singletons where the RHS is just a regular Rust expression. So for wrapping a value in a `RefCell` for example we could have a syntax like `#wrapped = RefCell::new(#unwrapped_singleton)`

**Comment by MingweiSamuel** [comment]
Basic implementation, just for a `state_ref()` proof-of-concept, released in 0.6.1
